lib/internal/modules/esm/get_file_system.js

```diff
@@ -0,0 +1,69 @@
+'use strict';
+
+const {
+  FunctionPrototypeBind,
+  ObjectCreate,
+  ObjectKeys,
+  SafeMap,
+} = primordials;
+
+const realpathCache = new SafeMap();
+
+const internalFS = require('internal/fs/utils');
+const fs = require('fs');
+const fsPromises = require('internal/fs/promises').exports;
+const packageJsonReader = require('internal/modules/package_json_reader');
+const { fileURLToPath } = require('url');
+const { internalModuleStat } = internalBinding('fs');
+
+const implementation = {
+  async readFile(p) {
+    return fsPromises.readFile(p);
+  },
+
+  async statEntry(p) {
+    return internalModuleStat(fileURLToPath(p));
+  },
+
+  async readJson(p) {
+    return packageJsonReader.read(fileURLToPath(p));
+  },
+
+  async realpath(p) {
+    return fsPromises.realpath(p, {
+      [internalFS.realpathCacheKey]: realpathCache
+    });
+  },
+
+  readFileSync(p) {
+    return fs.readFileSync(p);
+  },
+
+  statEntrySync(p) {
+    return internalModuleStat(fileURLToPath(p));
+  },
+
+  readJsonSync(p) {
+    return packageJsonReader.read(fileURLToPath(p));
+  },
```

Comment:
Is there a reason we can't just have `readJson` as a well-defined abstraction on top of `readFile`? Package virtualization can still work by returning serialized JSON - which is likely needed anyway with threading workflows.

Reply:
I didn't make perf analysis on `internalReadJson`, but I'd intuitively agree that a JS abstraction over `readFile` (that would just apply a regex to get the `containsKey` value) would fast enough without requiring a dedicated native function. I can experiment with that. Or do you have something else in mind?